### PR TITLE
#36336 Prefix all Indexes and Constraints

### LINKED ISSUE #36336: add UNIQUE or UNQ name to unique indexes in database tables
### Description (*)

This table `eav_attribute_label` has an index defined in `db_schema.xml`

```
    <index referenceId="EAV_ATTRIBUTE_LABEL_ATTRIBUTE_ID_STORE_ID" indexType="btree">
        <column name="attribute_id"/>
        <column name="store_id"/>
    </index>
```

I work on fast database import scripts and tried to add a new unique index in that table:
```
    <table name="eav_attribute_label">
        <constraint xsi:type="unique" referenceId="EAV_ATTRIBUTE_LABEL_ATTRIBUTE_ID_STORE_ID_UNQ">
            <column name="attribute_id"/>
            <column name="store_id"/>
        </constraint>
    </table>
```

but unique index name create by Magento 2 is: `EAV_ATTRIBUTE_LABEL_ATTRIBUTE_ID_STORE_ID`
which is identical to the index defined above and that breaks installation / running `db:upgrade` schema

### Expected behavior (*)
Unique indexes should contain UNQ / UNIQUE in referenceId

### Benefits

No conflicts on `core` tables when new indexes or unique indexes are added.
Ability to create multiple insert scripts.

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
Pull Request enforces Magento to prefix MySQL object names, not only hash.
This change answers a need of having multiple objects inside of a single Table (e.g. Constraint and Index).

The change also simplifies the `trimHash` method to use hashes in a way known for Git commit hash (use the first 8 letters), instead of a method that... returns empty string.

Additionally, added the test coverage for "Just a hash" return. 

### Related Pull Requests
N/A

### Fixed Issues (if relevant)
1. Fixes magento/magento2#36336

### Manual testing scenarios (*)
1. Create two objects inside of the same table (`db_schema.xml`)
 ```xml
 <table name="quote_item">
   <constraint xsi:type="unique" referenceId="ThisDoesNotMatter">
     <column name="quote_item_id"/>
   </constraint>
   <index referenceId="ThisDoesNotMatterToo" indexType="btree">
     <column name="quote_item_id"/>
   </index>
 </table>
 ```
2. Execute `bin/magento setup:upgrade`

### Current Behavior
```
SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'XXX_QUOTE_ITEM_QUOTE_ITEM_ID', query was: ALTER TABLE `xxx_quote_item` ADD INDEX `XXX_QUOTE_ITEM_QUOTE_ITEM_ID` (`quote_item_id`)
```

### Expected Behavior

Execution `setup:upgrade` should pass correctly.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magentu run all tests

**Comment by lbajsarowicz** [comment]
@magento run Unit Tests

**Comment by lbajsarowicz** [comment]
@magento run Unit Tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

By the way, I tried to run Unit Tests locally, but... ~5% of them fail with whitespace mismatch:
```
13) Magento\Framework\Stdlib\Test\Unit\DateTime\DateTimeFormatterTest::testFormatObjectNumericFormat with data set #4 (1, 'March 30, 2022 at 12:01:02 AM...n Time')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'March 30, 2022 at 12:01:02 AM Greenwich Mean Time'
+'March 30, 2022 at 12:01:02 AM Greenwich Mean Time'

/var/www/html/lib/internal/Magento/Framework/Stdlib/Test/Unit/DateTime/DateTimeFormatterTest.php:165

14) Magento\Framework\Stdlib\Test\Unit\DateTime\DateTimeFormatterTest::testFormatObjectNumericFormat with data set #5 (0, 'Wednesday, March 30, 2022 at ...n Time')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Wednesday, March 30, 2022 at 12:01:02 AM Greenwich Mean Time'
+'Wednesday, March 30, 2022 at 12:01:02 AM Greenwich Mean Time'

/var/www/html/lib/internal/Magento/Framework/Stdlib/Test/Unit/DateTime/DateTimeFormatterTest.php:165

15) Magento\Framework\Stdlib\Test\Unit\DateTime\TimezoneTest::testGetDatetimeFormat with data set #0 ('en_US', 3, 'M/d/yy h:mm a')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'M/d/yy h:mm a'
+'M/d/yy h:mm a'
```

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @lbajsarowicz,

Thanks for the contribution!

The changes look good to me. Please fix the failed tests.

Thanks

**Comment by engcom-Dash** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/DB/ExpressionConverter.php
Comment (engcom-Hotel) [inline lib/internal/Magento/Framework/DB/ExpressionConverter.php]: `trimHash` method call is provided 3 parameters, but the original method signature uses 1 parameter.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/DB/ExpressionConverter.php
Comment (engcom-Hotel) [reply lib/internal/Magento/Framework/DB/ExpressionConverter.php]: Also suggesting you introduce a new method instead of making changes in the existing method `trimHash`, as removing existing arguments from the method might be backward incompatible. 

Also, mark `trimHash` method deprecate.

**Comment by engcom-Dash** [comment]
Hello @lbajsarowicz,

Thanks for the contribution!

I'm getting an open search error with your PR code. When I revert it, it works as expected. Could you please check it?

To reproduce the issue, run the `bin/magento set:up` command.

Please see the screenshot below.
image

File: lib/internal/Magento/Framework/DB/ExpressionConverter.php
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/DB/ExpressionConverter.php]: This is a `private` method (`trimHash`) so it does not need to be marked as deprecated.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
As discussed with @lbajsarowicz over Slack, he mentioned that he will address this [comment](https://github.com/magento/magento2/pull/37975#issuecomment-2373370709).

**Comment by engcom-Dash** [comment]
Closing this PR since it has not been updated for a period of 14 days. Please feel free to reopen it if needed.

**Comment by lbajsarowicz** [comment]
C'mon. People have work here, enough billable work to queue non-billable Test Coverage for later. 😠

**Comment by engcom-Dash** [comment]
Hello @lbajsarowicz,

Thank you for your contributions.

I am reopening this PR. Please address this comment, as it is blocking us from proceeding further: [#37975 (comment)](https://github.com/magento/magento2/pull/37975#issuecomment-2373370709).

Thanks!

**Comment by engcom-Dash** [comment]
Hello @lbajsarowicz,

Thank you for your contributions!

We are putting this PR on hold for now. Once the issue referenced in https://github.com/magento/magento2/pull/37975#issuecomment-2373370709 has been resolved, we’ll proceed with the next steps.

Thank you for your understanding!

**Comment by engcom-Dash** [comment]
Hi @lbajsarowicz,

This is a reminder to please address this https://github.com/magento/magento2/pull/37975#issuecomment-2373370709 so that we can proceed further with this.

Thank you!

**Comment by lbajsarowicz** [comment]
**Update** Working on it.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@engcom-Dash Fixed. The root cause was unexpected prefix added to Table Names in `\Magento\Framework\DB\Adapter\Pdo\Mysql::getTableName`.

@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Hello @lbajsarowicz,

Since its related issue has been marked as a `Feature Request`, we need a PO approval to move forward for this. Please allow us some time for the same.

Till then we are moving this PR `On Hold`.

Thanks for the contribution.

**Comment by engcom-Hotel** [comment]
Hello @lbajsarowicz,

We have received a response from the PO, indicating that renaming database table columns, indexes, or constraints is considered a breaking change. This could disrupt extensions and custom scripts that reference the original names. To adopt such changes, customers would need to implement database migration scripts to maintain data integrity and consistency, which carries a risk of data loss.

Given these implications, I believe it would be best not to proceed with this PR.

However, we sincerely appreciate your valuable contributions.

Thanks

**Pull request closed without merging**
